Adds continuous archiving feature.

isundaylee opened the pull request:
Implements `postgres:enable-wal <name>` that enables continuous archiving of WAL segments. This is useful as it allows incremental backup of the cluster (see http://www.postgresql.org/docs/9.1/static/continuous-archiving.html).

Archived WAL segments are written to $SERVICE_ROOT/backups, which is mounted in the container as /var/lib/postgresql/backups. User can then backup this directory on the host using traditional backup mechanisms (rsync, upload to S3, etc).

This is a feature that i have found useful in my backup workflow. Just putting it here to gauge interests of including it in the main dokku-postgres plugin. If people think this might be useful, I can polish it up, write test cases, and probably also write a command that allows streamlined recovery from archived WAL segments. 

Edit: Recover command is now implemented. See commit message of 4f57281.

josegonzalez commented:
Is there a way to "disable-wal"?

josegonzalez commented:
Also, needs docs changes.

isundaylee commented:
Yeah, I think disable should just revert the config? And shouldn't remove previous archive right? 

Also, I'm just putting it here to gauge interest. If you think it's useful, I will start documenting the feature. Is there any other documentation than the README file?

josegonzalez commented:
I think it's useful, though won't give it a thorough review until you document it.

Regarding reverting config, sounds good. Same with not removing previous archives, though that should be documented.

shulcsm commented:
I'm interested in this.

isundaylee commented:
Sorry for leaving this pull request hanging for a long time. Seeing that there's interest, I can take point on rebasing this on top of the newest master. Here are some of the work remaining to be done that I can think of: 

1) Rebasing.
2) Writing proper documentation. 
3) Adding test cases in the `tests` directory. 

Did I miss anything?

isundaylee commented:
Progress update: I rebased the PR on top of the newest master, and implemented `wal-disable`. And I am in the process of making documentation changes and writing tests. 

A question about the user interface: Currently, backup files are stored in `/var/lib/dokku/services/postgres/<db_name>/backups`. However, the directory and the files in it are not readable by a normal user. Any ideas on how we should expose that directory to users, so that they can actually back it up?

josegonzalez commented:
What user are they owned by?

isundaylee commented:
On my test machine it's owned by the `shiny` user -- same as the owner of the `data` directory.

josegonzalez commented:
I would think all of the files should be owned by the Dokku user?

isundaylee commented:
On my system all the data files are for some reason owned by the `shiny` user and the `docker` group. However, in all cases, the owner of the backup files are going to be the same as the owner of the data files. In that case, any idea on how we could expose that directory? In my setup I just use sudo when I back up those files. But I guess that would not be ideal.

isundaylee commented:
Update on this: Reason I have not been updating this is that in my testing on one production host, it turned out that taking 1-minute snapshots makes it very slow to restore from the snapshots (since the recovery process goes through the snapshots one-by-one). Therefore there needs to be a way to periodically squash old snapshots together. However I have not been able to find one yet.

Will self-close this for now.
